docs/maintaining/testing.rst

```diff
@@ -0,0 +1,57 @@
+Testing
+=======
+
+When making a contribution to opsdroid, it's important that tests are written to validate that the functionality works, and continues to work in the future.
+
+opsdroid is currently slowly migrating from `unittest <https://docs.python.org/3.8/library/unittest.html>`__ to `pytest <https://docs.pytest.org>`__ as it's testing framework.
+As a part of this migration tests can be found in one of two places, old unittest tests are found in the ``tests/`` directory in the root of the repository and new pytest tests are found in a ``tests/`` directory in the same subfolder as the code, i.e ``opsdroid/tests/``.
+
+
+Writing new tests with pytest
+-----------------------------
+
+As well as core pytest, opsdroid is making use of a few different pytest plugins:
+
+* `pytest-asyncio <https://github.com/pytest-dev/pytest-asyncio>`__ for writing and running coroutines as tests. Coroutines should be decorated with ``@pytest.mark.asyncio``.
+* `pytest-mock <https://github.com/pytest-dev/pytest-mock/>`__ for proving a convenient interface to mocking parts of the code.
+  This package is configured to use the external `mock <https://pypi.org/project/mock/>`__ package, so all versions of Python can use the latest features.
+
+
+An example test
+###############
+
+Below is an example test, using the pytest-mock, pytest-asyncio and a `fixture <https://docs.pytest.org/en/stable/fixture.html>`__.
+
+.. code-block:: python
+
+    from opsdroid.memory import Memory
+    from opsdroid.database import InMemoryDatabase
+    
+    
+    @pytest.fixture
+    def memory():
+        mem = Memory()
+        mem.databases = [InMemoryDatabase()]
+        return mem
+    
+    
+    @pytest.mark.asyncio
+    async def test_database_callouts(mocker, memory):
+        memory.databases = [mocker.AsyncMock()]
+        data = "Hello world!"
+    
+        await memory.put("test", data)
+        assert memory.databases[0].put.called
+
+A pytest fixture sets up an object once per test, making sure that the state is clean for every test that uses it.
+It can also perform teardown actions after a ``yield`` statement.
+
+The ``mocker`` fixture is provided by pytest-mock, and provides convenient access to things in the mock library, as well as automatic teardown of patches added with ``mocker.patch``.
+
+
+opsdroid test helpers (`opsdroid.testing`)
+------------------------------------------
+
+.. automodapi:: opsdroid.testing
+   :headings: -#
+   :no-heading:
```

Comment:
Is this the reason why you had to use RST here? 

I had a read of the docs for `automodapi` but struggled to see what it does differently to automodule.

I'd also like to understand more about why this wouldn't work inside an ` ```eval_rst``` ` block in markdown?

Reply:
automodapi generates a complete list of all the functions, classes and variables and generates all the API documentation for all of the objects in the module. `.. automodule` only captures the module docstring.

it appears to fail in an `eval_rst` because of the lack of interlinking support, as the main include works but none of the links to the object documentation work.

Reply:
Ah ok. So kinda like the option to automatically include methods when using `autoclass`.

I'm surprised by this. `eval_rst` within a markdown document and a regular RST file should both result in the same AST as they use the same parser. Then the same docs should be generated.

I've not seen issues like this with `autodoc` which I assume would suffer from the same problems.

Reply:
I didn't fall deep down the pit of working it out, I swapped it to rst to see if it fixed it and it did, then I pushed it ;)

Reply:
Looking at the generated source, something must be funky, because the HTML isn't generating a link. So it's not that it's generating a broken link, sphinx isn't generating the link at all...